### PR TITLE
Metadata Plugin API

### PR DESCRIPTION
**Pull request by rohitsinha54**
- This PR exposes the MetadataReaderContext and WriterContext to PluginContext allowing user to use these API to do metadata operation.
- There is no implementation to the apis as of now and the purpose is to just show how the api look like and how the user will use it.
- There is a sample example of how the api will be used but as obvious it does nothing actually.

### CODE REVIEW

File: cdap-app-templates/cdap-data-quality/src/test/java/co/cask/cdap/dq/testclasses/StreamBatchSource.java [hunk 1]
```diff
@@ -115,6 +120,37 @@ public void prepareRun(BatchSourceContext context) {
...
+    propogateMetadata(context, sourceEntity, sinkEntity);
+
+    // Propagate metadata of some fields
+    MetadataEntity sourceFieldEntity = sourceEntity.append("field", "empName");
+    MetadataEntity sinkFieldEntity = sinkEntity.append("field", "empName");
+    propogateMetadata(context, sourceFieldEntity, sinkFieldEntity);
+
+    for (Schema.Field field : context.getInputSchema().getFields()) {
+      sourceFieldEntity = sourceEntity.append("field", field.getName());
+      Metadata metadata = context.getMetadata(MetadataScope.USER, sourceFieldEntity);
```
Comment (anew) [inline cdap-app-templates/cdap-data-quality/src/test/java/co/cask/cdap/dq/testclasses/StreamBatchSource.java:141]: Should it be necessary to specify Scope.USER? Or isn't it true that the user can only manipulate USER metadata? What is an example of Scope.SYSTEM metadata?

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/TransformContext.java [hunk 2]
```diff
@@ -17,11 +17,13 @@
...
 import co.cask.cdap.api.annotation.Beta;
+import co.cask.cdap.api.metadata.MetadataReaderContext;
+import co.cask.cdap.api.metadata.MetadataWriterContext;
 
 /**
  * Context passed to ETL Transform stages.
  */
 @Beta
-public interface TransformContext extends StageContext, LookupProvider {
+public interface TransformContext extends StageContext, LookupProvider, MetadataReaderContext, MetadataWriterContext {
```
Comment (anew) [inline cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/TransformContext.java]: Should this be done for all types of stages? That is, StageContext would extend the two new interfaces. Or how can an action plugin or a error collector manipulate metadata? Do both sources and sinks have a transform context?

File: cdap-app-templates/cdap-data-quality/src/test/java/co/cask/cdap/dq/testclasses/StreamBatchSource.java (hunk 1 above)
Comment (rohitsinha54) [reply cdap-app-templates/cdap-data-quality/src/test/java/co/cask/cdap/dq/testclasses/StreamBatchSource.java:141]: User's can get both system and/or user metadata although they can only add/manipulate metadata in User scope. There is another api which does not take a scope and return metadata for both the scopes.

Comment (rohitsinha54) [reply cdap-app-templates/cdap-data-quality/src/test/java/co/cask/cdap/dq/testclasses/StreamBatchSource.java:141]: I have moved them to StageContext as needed later on during the implementation we can add implementations/overrides wherever needed.

**Review by anew — APPROVED** [review]

**Comment by rohitsinha54** [comment]
Build: https://builds.cask.co/browse/CDAP-DUT6447-1

**Comment by rohitsinha54** [comment]
Build: https://builds.cask.co/browse/CDAP-DUT6469-1

**Pull request merged**